### PR TITLE
Implemented Flowfield Pathfinding

### PR DESCRIPTION
**Pull request by sarimmehdi**
This algorithm is used in real-time strategy games to do unit pathfinding.

### CODE REVIEW

File: PathPlanning/FlowField/flowfield.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    def find_path(self):
+        self.create_cost_field()
+        self.create_integration_field()
+        self.assign_vectors()
+        self.follow_vectors()
+
+    def create_cost_field(self):
+        """Assign cost to each grid which defines the energy
+        it would take to get there."""
+        self.cost_field = {}
```
Comment (AtsushiSakai) [inline PathPlanning/FlowField/flowfield.py:46]: This field should be initialized in the __init__()

File: PathPlanning/FlowField/flowfield.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                    if neighbor_new_cost < neighbor_old_cost:
+                        self.integration_field[(x, y)] = neighbor_new_cost
+                        open_list.append(([x, y], neighbor_new_cost))
+            del open_list[0]
+
+    def assign_vectors(self):
+        """For each node, assign a vector from itself to the node with
+        the lowest cost in the integration field. An agent will simply
+        follow this vector field to the goal"""
+        self.vector_field = {}
```
Comment (AtsushiSakai) [inline PathPlanning/FlowField/flowfield.py:110]: This field should be initialized in the __init__()

File: PathPlanning/FlowField/flowfield.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        (must not be obstacles). For each neighbor, the new
+        cost is equal to the cost of the current node in the
+        integration field (in the beginning, this will simply
+        be the goal node) + the cost of the neighbor in the
+        cost field + the extra cost (optional). The new cost
+        is only assigned if it is less than the previously
+        assigned cost of the node in the integration field and,
+        when that happens, the neighbor is put on the open list.
+        This process continues until the open list is empty."""
+        self.integration_field = {}
```
Comment (AtsushiSakai) [inline PathPlanning/FlowField/flowfield.py:76]: This field should be initialized in the __init__()

File: PathPlanning/FlowField/flowfield.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+    for x, y, l in zip(all_x, all_y, all_len):
+        draw_horizontal_line(x, y, l, o_x, o_y, obs_dict, 'obs')
+
+    # draw paths that have medium difficulty (in terms of going through them)
+    all_x[:], all_y[:], all_len[:] = [], [], []
+    all_x = [10, 45]
+    all_y = [22, 20]
+    all_len = [8, 5]
+    for x, y, l in zip(all_x, all_y, all_len):
+        draw_vertical_line(x, y, l, m_x, m_y, obs_dict, 'medium')
```
Comment (AtsushiSakai) [inline PathPlanning/FlowField/flowfield.py:178]: What do you mean the "medium" and "hard"? Could you please add comment to explain it in here or header comment?

File: PathPlanning/FlowField/flowfield.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+    plt.plot(o_x, o_y, "sr")
+    plt.plot(m_x, m_y, "sg")
+    plt.plot(h_x, h_y, "sy")
+    plt.plot(s_x, s_y, "og")
+    plt.plot(g_x, g_y, "o")
+    plt.grid(True)
+
+    flow_obj = FlowField(obs_dict, g_x, g_y, s_x, s_y, 50, 50)
+    flow_obj.find_path()
```
Comment (AtsushiSakai) [inline PathPlanning/FlowField/flowfield.py:210]: What is these color (yellow, green) meaning? Could you please add description to explain it.
SS 2020-09-27 at 13 37 36

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Thank you for interesting PR. I have some comments. PTAL.

**Comment by AtsushiSakai** [comment]
One more thing, from the next PR. Could you please change branch for each PR? It is better for checking the git log in PR thread. You don't need to do anything in this PR.

**Comment by sarimmehdi** [comment]
@AtsushiSakai I have made the requested changes. Please have a look.

File: PathPlanning/FlowField/flowfield.py [hunk 6]
```diff
@@ -169,6 +169,15 @@ def main():
     for x, y, l in zip(all_x, all_y, all_len):
         draw_horizontal_line(x, y, l, o_x, o_y, obs_dict, 'obs')
 
+    # Some points are asigned a slightly higher energy value in the cost
```
Comment (AtsushiSakai) [inline PathPlanning/FlowField/flowfield.py:172]: asigned -> assigned ?

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Thank you for fixing. I have just one small comment. After fixing it, I will merge it.

File: PathPlanning/FlowField/flowfield.py (hunk 6 above)
Comment (sarimmehdi) [reply PathPlanning/FlowField/flowfield.py]: I have made the change.

**Review by AtsushiSakai — APPROVED** [review]
LGTM. Thanks for your contribution!!

**Pull request merged**